### PR TITLE
IGNITE-15272 Async raft group service startup

### PR DESCRIPTION
**Pull request by SammyVimes**

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java [hunk 1]
```diff
@@ -117,6 +122,14 @@ public ConnectionManager(
      */
     public void start() throws IgniteInternalException {
         try {
+            boolean wasStarted = started.getAndSet(true);
```
Comment (ibessonov) [inline modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java:125]: Have we discussed guarantees that start/stop should provide when called in parallel? Right now I can see a race

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java [hunk 2]
```diff
@@ -232,11 +245,15 @@ public void addListener(BiConsumer<SocketAddress, NetworkMessage> listener) {
...
+        boolean wasStopped = this.stopped.getAndSet(true);
+
+        if (wasStopped)
+            return;
+
+        Stream<CompletableFuture<Void>> stream = Stream.concat(
             clients.values().stream().map(NettyClient::stop),
             Stream.of(server.stop())
         );
-
```
Comment (ibessonov) [inline modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java:239]: Can you please put it back?

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java [hunk 3]
```diff
@@ -155,16 +187,48 @@ public RaftGroupServiceImpl(
...
+        );
+
+        return fut.thenApply(GetLeaderResponse::leader);
+    }
 
-        return fut.thenApply(resp -> {
-            leader = resp.leader();
+    /** {@inheritDoc} */
+    @Override public CompletableFuture<Void> refreshLeader() {
+        return getLeader(groupId, factory, cluster, peers, timeout, retryDelay).thenApply(leader -> {
```
Comment (ibessonov) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java:230]: Does "thenAccept" fit better here?

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java [hunk 4]
```diff
@@ -155,16 +187,48 @@ public RaftGroupServiceImpl(
...
+     *
+     * @param groupId Raft group id.
+     * @param factory Message factory.
+     * @param cluster Cluster service.
+     * @param peers List of all peers.
+     * @param timeout Timeout.
+     * @param retryDelay Retry delay.
+     * @return Future representing pending completion of the request.
+     */
+    private static CompletableFuture<Peer> getLeader(
```
Comment (ibessonov) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java:201]: Can you make it not static and invoke it on new instance before completing the future in "start"? Too many arguments

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java [hunk 5]
```diff
@@ -88,24 +88,26 @@
     private final long retryDelay;
 
     /** */
-    private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
+    private static final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java:91]: well, if every raft group has its own scheduler, we end up with 1024 schedulers for one table. We can live it as it is for now, but it sure needs some refactoring

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/server/ITSimpleCounterServerTest.java [hunk 6]
```diff
@@ -103,39 +105,29 @@ void before(TestInfo testInfo) {
...
         assertTrue(waitForTopology(clientNode2, 2, 1000));
     }
 
     /**
-     * @throws Exception
+     * @throws Exception If failed.
      */
     @AfterEach
-    void after() throws Exception {
+    @Override public void after(TestInfo testInfo) throws Exception {
```
Comment (ibessonov) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/server/ITSimpleCounterServerTest.java:125]: Why do you override "after" but not "before"?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 7]
```diff
@@ -196,21 +196,35 @@ private void createTableLocally(
...
+            partitionsGroupsFutures.add(raftMgr.prepareRaftGroup(
                 raftGroupName(tblId, p),
                 assignment.get(p),
                 new PartitionListener(storage)
             ));
         }
 
-        InternalTableImpl internalTable = new InternalTableImpl(name, tblId, partitionMap, partitions);
+        CompletableFuture.allOf(partitionsGroupsFutures.toArray(CompletableFuture[]::new)).thenRun(() -> {
+            HashMap<Integer, RaftGroupService> partitionMap = new HashMap<>(partitions);
```
Comment (ibessonov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:207]: You replaced map with list in code above but not here. Why?

**Pull request merged**